extensions-core/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesClient.java

```diff
@@ -19,21 +19,64 @@
 
 package org.apache.druid.k8s.overlord.common;
 
+import com.google.common.base.Preconditions;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.batch.v1.Job;
 import io.fabric8.kubernetes.client.Config;
 import io.fabric8.kubernetes.client.KubernetesClient;
 import io.fabric8.kubernetes.client.KubernetesClientBuilder;
+import io.fabric8.kubernetes.client.informers.ResourceEventHandler;
+import io.fabric8.kubernetes.client.informers.SharedIndexInformer;
+import org.apache.druid.error.DruidException;
+import org.apache.druid.java.util.emitter.EmittingLogger;
 import org.apache.druid.k8s.overlord.common.httpclient.DruidKubernetesHttpClientFactory;
 
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
 public class DruidKubernetesClient implements KubernetesClientApi
```

Comment:
All the changes in this class (and the corresponding APIs in `KubernetesClientApi`) pertain only to the caching implementation.

I would suggest skipping the changes made to `KubernetesClientApi` and `DruidKubernetesClient` and move all the new logic into a separate class altogether and use it in `CachingKubernetesPeonClient` directly. The non-caching impl shouldn't not even be able to invoke these (rather than initializing podInformer etc to null and throw exceptions if invoked).

The new class may use `DruidKubernetesClient.getClient()` if we want to avoid creating a fresh client.